src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,510 @@
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Extensions;
+
+[ExportWorkspaceServiceFactory(typeof(IExtensionMessageHandlerService)), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class ExtensionMessageHandlerServiceFactory(IExtensionMessageHandlerFactory customMessageHandlerFactory)
+    : IWorkspaceServiceFactory
+{
+    private readonly record struct AssemblyMessageHandlers(
+        ImmutableDictionary<string, IExtensionMessageHandlerWrapper<Document>> DocumentMessageHandlers,
+        ImmutableDictionary<string, IExtensionMessageHandlerWrapper<Solution>> WorkspaceMessageHandlers);
+
+    public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        => new ExtensionMessageHandlerService(
+            workspaceServices.SolutionServices,
+            customMessageHandlerFactory);
+
+    private sealed class ExtensionMessageHandlerService(
+        SolutionServices solutionServices,
+        IExtensionMessageHandlerFactory customMessageHandlerFactory)
+        : IExtensionMessageHandlerService
+    {
+        private static readonly ConditionalWeakTable<IExtensionMessageHandlerWrapper, IExtensionMessageHandlerWrapper> s_disabledExtensionHandlers = new();
+
+        private readonly SolutionServices _solutionServices = solutionServices;
+        private readonly IExtensionMessageHandlerFactory _customMessageHandlerFactory = customMessageHandlerFactory;
+
+        /// <summary>
+        /// Lock for <see cref="_folderPathToExtensionFolder"/>, <see cref="_cachedDocumentHandlers"/>, and <see
+        /// cref="_cachedWorkspaceHandlers"/>.  Note: this type is designed such that all time while this lock is held
+        /// should be minimal.  Importantly, no async work or IO should be done while holding this lock.  Instead,
+        /// all of that work should be pushed into AsyncLazy values that compute when asked, outside of this lock.
+        /// </summary>
+        private readonly object _gate = new();
+
+        /// <summary>
+        /// Extensions assembly load contexts and loaded handlers, indexed by extension folder path.
+        /// </summary>
+        private readonly Dictionary<string, ExtensionFolder> _folderPathToExtensionFolder = new();
+
+        /// <summary>
+        /// Cached handlers of document-related messages, indexed by handler message name.
+        /// </summary>
+        private readonly Dictionary<string, AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper<Document>>>> _cachedDocumentHandlers = new();
+
+        /// <summary>
+        /// Cached handlers of non-document-related messages, indexed by handler message name.
+        /// </summary>
+        private readonly Dictionary<string, AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper<Solution>>>> _cachedWorkspaceHandlers = new();
```

Comment:
to make things simpler, instead of trying to populate these up front, they are populated on demand, and cleared when extensions are registered/unregistered.